### PR TITLE
refactor(tcp): use SelectAll for driving listener streams

### LINKED ISSUE #2781: transports/tcp: use `futures::stream::SelectAll` for driving listener streams
## Description

With #2652 polling the listeners of a transport became the transport's responsibility. 
We currently do this in `TcpTransport::poll` analogous to how it was done before in the swarm `ListenerStream`: by iterating through the list of `TcpListenStream`s and sequentially polling each listener for an event, returning the first event from a listener.
Instead of doing this manual iteration, we should use [`futures::stream::SelectAll`](https://docs.rs/futures/latest/futures/stream/select_all/struct.SelectAll.html#) for polling all `TcpListenStream`s. This is already used in e.g. the relay [`ClientTransport`](https://github.com/libp2p/rust-libp2p/blob/62622a1bad70381e56237b4da4d9abcf2b1a0e24/protocols/relay/src/v2/client/transport.rs#L89) and the [`wasm_ext::ExtTransport`](https://github.com/libp2p/rust-libp2p/blob/62622a1bad70381e56237b4da4d9abcf2b1a0e24/transports/wasm-ext/src/lib.rs#L148).

The following changes would be required:
- Remove `TcpListenerEvent`, instead `<TcpListenStream as Stream>::Item` should directly return a `TransportEvent`.
- When a listener was closed, `<TcpListenStream as Stream>::poll_next` should return `Poll::Ready(None)` for this listener so that the `SelectAll` combinator drops the stream
- Manage the `TcpListenStream`s in a `SelectAll`, poll the `SelectAll` stream in `TcpTransport::poll`.

## Motivation

`SelectAll` is much more efficient when driving multiple streams since it only polls its inner streams when they generate notifications.
Apart from that, it reduces the complexity as we only have to poll the `SelectAll` stream rather than then manual iteration and polling.

## Current Implementation

The `TcpListenStream`s are pinned in a `VecDeque`. When the task is woken and `TcpTransport::poll` called, we iterate through this list by popping the first element, polling it for new events, pushing the listener back onto the list. If the listener returned an `TcpListenerEvent` we return and `TransportEvent` for it, else the next listener is polled:
https://github.com/libp2p/rust-libp2p/blob/eaf3f3a7fb4aba228ebd15366ab79b64cf37832c/transports/tcp/src/lib.rs#L509-L582

## Are you planning to do it yourself in a pull request?

No (or at least not in the near future).

### PR DESCRIPTION
**Pull request by vnermolaev**
## Description

The PR optimizes polling of the listeners in the TCP transport by using `futures::SelectAll` instead of storing them in a queue and polling manually.

Resolves #2781.

## Notes

I observed that the following branches are never triggered
https://github.com/libp2p/rust-libp2p/blob/d82c2a1a7a4032a4715bd079043a7609ef56080e/transports/tcp/src/lib.rs#L600-L611

This is because the underlying [`ListenStream`](https://github.com/libp2p/rust-libp2p/blob/d82c2a1a7a4032a4715bd079043a7609ef56080e/transports/tcp/src/lib.rs#L642) always returns `Poll::Ready(Some(Ok(..)))` and neither `Poll::Ready(Some(Err(...)))` (in-line documentation says such errors are not fatal), nor `Poll::Ready(None)`.

I ended up following the approach found in  

https://github.com/vnermolaev/rust-libp2p/blob/acf84113cff5843234ba6ea20c79b079e9d1195c/protocols/relay/src/priv_client/transport.rs#L232-L235

Please advise. Is this the desired behaviour? Otherwise, what approach should I follow? @elenaf9 

## Links to any relevant issues

#2781

## Change checklist

- [x] I have performed a self-review of my own code
- [x] I have made corresponding changes to the documentation
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] A changelog entry has been made in the appropriate crates

### CODE REVIEW

File: transports/tcp/src/lib.rs
Comment (jxs) [inline transports/tcp/src/lib.rs]:
```suggestion
```

File: protocols/relay/CHANGELOG.md
Comment (jxs) [inline protocols/relay/CHANGELOG.md]:
```suggestion
- Optimize listeners polling behaviour in the tcp transport, fixing [2781](https://github.com/libp2p/rust-libp2p/issues/2781). See [PR 3361]
```

File: protocols/relay/CHANGELOG.md
Comment (jxs) [inline protocols/relay/CHANGELOG.md]:
```suggestion
[PR 3361]: https://github.com/libp2p/rust-libp2p/pull/3361

```

File: transports/tcp/src/lib.rs
Comment (jxs) [inline transports/tcp/src/lib.rs:456]: Shouldn't we continue to remove the listener?

**Review by jxs — COMMENTED** [review]
Hi, and thanks for tackling this! Some notes left

File: transports/tcp/src/lib.rs
Comment (vnermolaev) [reply transports/tcp/src/lib.rs]: Addressed

File: protocols/relay/CHANGELOG.md
Comment (vnermolaev) [reply protocols/relay/CHANGELOG.md]: Addressed

File: protocols/relay/CHANGELOG.md
Comment (vnermolaev) [reply protocols/relay/CHANGELOG.md]: Addressed

File: transports/tcp/src/lib.rs
Comment (vnermolaev) [reply transports/tcp/src/lib.rs:456]: Indeed. This one was a bit trickier. I followed the approach laid out for https://github.com/libp2p/rust-libp2p/blob/d82c2a1a7a4032a4715bd079043a7609ef56080e/protocols/relay/src/priv_client/transport.rs#L300

File: transports/tcp/src/lib.rs
Comment (thomaseizinger) [inline transports/tcp/src/lib.rs]: We now have two return paths here that don't register their own waker. You'll need to store a waker in the listener before you return `Poll::Pending` at the bottom.

File: transports/tcp/src/lib.rs
Comment (thomaseizinger) [inline transports/tcp/src/lib.rs]: In case the waker is set, you need to call it here to make sure you are getting polled.

File: protocols/relay/CHANGELOG.md
Comment (thomaseizinger) [inline protocols/relay/CHANGELOG.md]: Why are we making a changelog entry for relay when we are changing code in TCP?

File: transports/tcp/src/lib.rs
Comment (thomaseizinger) [inline transports/tcp/src/lib.rs]: Can we merge these two `use` blocks please?

File: transports/tcp/src/lib.rs
Comment (thomaseizinger) [reply transports/tcp/src/lib.rs:456]: `SelectAll` doesn't allow you to remove elements. The `close` solution is good :)

File: transports/tcp/src/lib.rs
Comment (thomaseizinger) [inline transports/tcp/src/lib.rs]:
```suggestion
    /// The listener can be manually closed with [`Transport::remove_listener`](libp2p_core::Transport::remove_listener).
```

File: transports/tcp/src/lib.rs
Comment (thomaseizinger) [inline transports/tcp/src/lib.rs]: Do we ever push something else in here? If not, then a more specific solution would be:

```rust
enum State {
   Active,
   Closing,
   Closed
}
```

In `Closing` we can then report the `Closed` event, set ourselves to `Closed` and return `None`.

File: transports/tcp/src/lib.rs
Comment (thomaseizinger) [inline transports/tcp/src/lib.rs]: Remove these comments please.

File: protocols/relay/CHANGELOG.md
Comment (thomaseizinger) [reply protocols/relay/CHANGELOG.md]: Btw, I don't think this needs a changelog entry at all because it is an internal refactoring.

**Review by thomaseizinger — COMMENTED** [review]
A few comments, otherwise LGTM.

Clean work, thank you!

File: protocols/relay/CHANGELOG.md
Comment (vnermolaev) [reply protocols/relay/CHANGELOG.md]: Fixed. The PR templates explicitly asks to make a changelog entry. I am not making the rules :)

File: transports/tcp/src/lib.rs
Comment (vnermolaev) [reply transports/tcp/src/lib.rs]: fixed

File: transports/tcp/src/lib.rs
Comment (vnermolaev) [reply transports/tcp/src/lib.rs]: fixed

File: transports/tcp/src/lib.rs
Comment (vnermolaev) [reply transports/tcp/src/lib.rs]: fixed

File: protocols/relay/CHANGELOG.md
Comment (thomaseizinger) [reply protocols/relay/CHANGELOG.md]: Fair! Most PRs, especially from external contributors add functionality or fix bugs, both of which are user-facing. This one is the exception to the rule :)

File: transports/tcp/src/lib.rs
Comment (jxs) [inline transports/tcp/src/lib.rs]:
```suggestion
```

File: transports/tcp/src/lib.rs
Comment (vnermolaev) [reply transports/tcp/src/lib.rs]: I lack expertise on this particular issue. Instead I have aligned the impl for the `tcp` transport with the impl for the [`quic` transport](https://github.com/libp2p/rust-libp2p/blob/4b41f5a9942add3f7a815bd2b79cb82eaed1ee05/transports/quic/src/transport.rs#L316), they are pretty similar.

File: transports/tcp/src/lib.rs
Comment (vnermolaev) [reply transports/tcp/src/lib.rs]: It is now an `Option`. It is aligned with at least `quic` and `webrtc`; I did not see such an approach with enums in other transports...

File: transports/tcp/src/lib.rs
Comment (thomaseizinger) [reply transports/tcp/src/lib.rs]: @elenaf9 Did you consider this case when implementing QUIC? I think we may have a potential unwoken task here:

- Assume 1 active listener
- Assume the listener had no work and returned `Pending`
- We call `close` on the listener
- Without an explicit waker, I believe the close will only happen once the listener gets polled again due to another event

Comment (vnermolaev) [reply transports/tcp/src/lib.rs]: Well, it appears this PR can't be merged before this question is cleared out.

Comment (thomaseizinger) [reply transports/tcp/src/lib.rs]: We can move forward here by just adding the waker. Worst case, we have an unnecessary wake-up.

- Make a new field inside the listener, holding an `Option<Waker>`
- Call it `close_listener_waker`
- Set it before you return `Poll::Pending` at the bottom
- Use `take` to remove the listener within `close` and if it is present, call `wake` on it

Comment (thomaseizinger) [reply transports/tcp/src/lib.rs]: @jxs @mxinden perhaps you want to take a look at this too.

File: transports/tcp/src/lib.rs
Comment (mxinden) [inline transports/tcp/src/lib.rs:551]: Great to see this going away. Thank you!

File: transports/tcp/src/lib.rs
Comment (mxinden) [reply transports/tcp/src/lib.rs]: In my eyes, the suggestion above by @thomaseizinger is the way to go.

**Comment by vnermolaev** [comment]
@mxinden @thomaseizinger 
I have applied the requested changes, i.e., 
> We can move forward here by just adding the waker. Worst case, we have an unnecessary wake-up.
> 
>     * Make a new field inside the listener, holding an `Option<Waker>`
> 
>     * Call it `close_listener_waker`
> 
>     * Set it before you return `Poll::Pending` at the bottom
> 
>     * Use `take` to remove the listener within `close` and if it is present, call `wake` on it

However, 
1. now, there are linting errors coming from parts other than the TCP transport. Should they be addressed separately?
Also the following lint
```bash
error: lint `clippy::derive_hash_xor_eq` has been renamed to `clippy::derived_hash_with_manual_eq`
Error:   --> muxers/mplex/src/codec.rs:68:14
   |
68 |     #![allow(clippy::derive_hash_xor_eq)]
   |              ^^^^^^^^^^^^^^^^^^^^^^^^^^ help: use the new name: `clippy::derived_hash_with_manual_eq`
   |
   = note: `-D renamed-and-removed-lints` implied by `-D warnings`
```
if applied will also break the build stating that such a lint is not known... can someone confirm?
2. The issue with sending the last event on closing and thus setting up a waker, I assume, also exists in other transports as it was the case and this one.

File: transports/tcp/src/lib.rs
Comment (elenaf9) [inline transports/tcp/src/lib.rs]:
```suggestion
    /// This will create a [`TransportEvent::ListenerClosed`] and
    /// terminate the stream once the event has been reported.
```

File: transports/tcp/src/lib.rs
Comment (elenaf9) [reply transports/tcp/src/lib.rs]: Good catch @thomaseizinger. Yes you are right, we need the explicit waker here.

**Review by elenaf9 — COMMENTED** [review]
Thank you for tackling this @vnermolaev! 

:+1: LGTM


> 1. now, there are linting errors coming from parts other than the TCP transport. Should they be addressed separately?

They are already addressed in #3389. Would you mind reverting ab06ff86ec8886f665b4d68d5d4b7d5230105353? 

> 2. The issue with sending the last event on closing and thus setting up a waker, I assume, also exists in other transports as it was the case and this one.

Yes you are right. Would you be interested in doing patches to fix them there as well (in separate PRs; namely for `libp2p-quic` and `libp2p-webrtc`)?

**Comment by vnermolaev** [comment]
> > 1. now, there are linting errors coming from parts other than the TCP transport. Should they be addressed separately?
> 
> They are already addressed in #3389. Would you mind reverting [ab06ff8](https://github.com/libp2p/rust-libp2p/commit/ab06ff86ec8886f665b4d68d5d4b7d5230105353)?
> 
@elenaf9  I have rebased on top of the latest master, excluding the mentioned commit, clippy still fails on sources outside the TCP transport, i.e., `swarm`, see the failing check.

**Comment by elenaf9** [comment]
> > > 1. now, there are linting errors coming from parts other than the TCP transport. Should they be addressed separately?
> > 
> > 
> > They are already addressed in #3389. Would you mind reverting [ab06ff8](https://github.com/libp2p/rust-libp2p/commit/ab06ff86ec8886f665b4d68d5d4b7d5230105353)?
> 
> @elenaf9 I have rebased on top of the latest master, excluding the mentioned commit, clippy still fails on sources outside the TCP transport, i.e., `swarm`, see the failing check.

Yes, #3389 is not merged yet. The failing check however is only the `clippy beta` one, which is not required for merging the PR. Ideally #3389 will be merged soon, but it shouldn't block this PR, i.e. we can also merge without.

**Comment by vnermolaev** [comment]
> Yes, #3389 is not merged yet. The failing check however is only the `clippy beta` one, which is not required for merging the PR. Ideally #3389 will be merged soon, but it shouldn't block this PR, i.e. we can also merge without.

then I think it is ready to be merged :)

> Yes you are right. Would you be interested in doing patches to fix them there as well (in separate PRs; namely for `libp2p-quic` and `libp2p-webrtc`)?

Yes, I am interested.

File: transports/tcp/CHANGELOG.md
Comment (thomaseizinger) [inline transports/tcp/CHANGELOG.md]: Do we need this? It is mostly a refactoring that is not user facing.

Could mention the bug fix but otherwise I'd suggest not adding a changelog entry.

File: transports/tcp/src/lib.rs
Comment (thomaseizinger) [inline transports/tcp/src/lib.rs]: Why are we deleting this test?

Comment (vnermolaev) [reply transports/tcp/src/lib.rs]: My bad, it's back.

File: transports/tcp/CHANGELOG.md
Comment (vnermolaev) [reply transports/tcp/CHANGELOG.md]: removed.

**Comment by vnermolaev** [comment]
All right, all suggestions/requests are satisfied.
Unfortunately, some CI checks fail, not triggered by changes in this PR.
```bash
Checking libp2p-request-response v0.24.0 (/home/runner/work/rust-libp2p/rust-libp2p/protocols/request-response)
error[E0369]: cannot multiply `&&EdwardsBasepointTable` by `&Scalar`
   --> /home/runner/.cargo/registry/src/github.com-1ecc6299db9ec823/snow-0.9.0/src/resolvers/default.rs:132:47
    |
132 |         let point = (&ED25519_BASEPOINT_TABLE * &self.privkey).to_montgomery();
    |                      ------------------------ ^ ------------- &Scalar
    |                      |
    |                      &&EdwardsBasepointTable
    |
help: `*` can be used on `&EdwardsBasepointTable` if you dereference the left-hand side
    |
132 |         let point = (*&ED25519_BASEPOINT_TABLE * &self.privkey).to_montgomery();
    |                      +
```
and clippy-beta as before.

**Comment by thomaseizinger** [comment]
See https://github.com/mcginty/snow/issues/146 for the build failure.

**Comment by thomaseizinger** [comment]
> See [mcginty/snow#146](https://github.com/mcginty/snow/issues/146) for the build failure.

I am afraid we cannot fix this on our end. We'd have to insert `cargo update` statements in our CI pipeline which doesn't help our users either. Let's see how responsive the `snow` maintainers are.

**Comment by vnermolaev** [comment]
Do I understand correctly that this PR is on hold for now?

**Comment by thomaseizinger** [comment]
> Do I understand correctly that this PR is on hold for now?

Yes, we can't merge PRs at the moment due to the breakage in `snow`.

I have a proposal up to track `Cargo.lock` in Git which allows us to pin an earlier version of `curve25519-dalek`: https://github.com/libp2p/rust-libp2p/pull/3399

**Review by thomaseizinger — DISMISSED** [review]

**Review by thomaseizinger — APPROVED** [review]

**Pull request merged**
